File: gtci/12-tree-depth-first-search/cpp/5-path-with-given-sequence.cpp
#include <cassert>
#include <iostream>
#include <vector>

#include "tree_utils.h"

using namespace std;

class Solution {
public:
  bool findPath(TreeNode* root, const vector<int>& sequence) {
    if (root == nullptr && sequence.empty())
      return true;

    std::vector<int> currPath;
    
    return findPathHelper(root, sequence, currPath);
  }

  bool findPathHelper(TreeNode* node, const vector<int>& sequence, vector<int>& currPath) {
    if (node == nullptr)
      return false;
    
    currPath.push_back(node->val);

    if (node->left == nullptr && node->right == nullptr && sequence == currPath) {
      currPath.pop_back();
      return true;
    }
    
    if (findPathHelper(node->left, sequence, currPath) || findPathHelper(node->right, sequence, currPath)) {
      currPath.pop_back();
      return true;
    }

    currPath.pop_back();
    return false;
  }
};

int main() {
    Solution sol;
    
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(0);
    root->right = new TreeNode(1);
    root->left->left = new TreeNode(1);
    root->right->left = new TreeNode(6);
    root->right->right = new TreeNode(5);
    
    vector<int> sequence = {1, 0, 7};
    assert(!sol.findPath(root, sequence));
    
    sequence = {1, 1, 6};
    assert(sol.findPath(root, sequence));
    
    sequence = {1, 0, 1};
    assert(sol.findPath(root, sequence));
    
    sequence = {1, 1, 1};
    assert(!sol.findPath(root, sequence));

    std::cout << "All test cases passed." << std::endl;

    delete root;
    
    return 0;
}

File: gtci/12-tree-depth-first-search/python3/5-path-with-given-sequence.py
class TreeNode:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def findPath(self, root, sequence):
        if root is None and not sequence:
            return True

        return self.findPath_helper(root, sequence, [])

    def findPath_helper(self, node, sequence, curr_path):
        if node is None:
            return False

        curr_path = curr_path + [node.val]

        if node.left is None and node.right is None:
            return curr_path == sequence

        return self.findPath_helper(
            node.left, sequence, curr_path
        ) or self.findPath_helper(node.right, sequence, curr_path)


if __name__ == "__main__":
    sol = Solution()

    root = TreeNode(1)
    root.left = TreeNode(7)
    root.right = TreeNode(9)
    root.right.left = TreeNode(2)
    root.right.right = TreeNode(9)
    assert sol.findPath(root, [1, 9, 9])

    root = TreeNode(1)
    root.left = TreeNode(0)
    root.right = TreeNode(1)
    root.left.left = TreeNode(1)
    root.right.left = TreeNode(6)
    root.right.right = TreeNode(5)
    assert not sol.findPath(root, [1, 0, 7])
    assert sol.findPath(root, [1, 1, 6])

    print("All test cases passed.")
